File: latest/week2/workshop6_m6/W06_M06_demo.R
# ============================================================
# Workshop 6 — Module 6: Animation Fundamentals
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── 1. SIMULATED GAPMINDER-STYLE DATA ───────────────────
set.seed(42); n_countries <- 40
base_gdp <- exp(runif(n_countries, 6, 10))
continents <- sample(c("Europe", "Asia", "Africa", "Americas"),
  n_countries, replace = TRUE, prob = c(0.25, 0.30, 0.25, 0.20))

df <- tibble()
for (yr in 2000:2020) {
  noise_gdp <- rnorm(n_countries, 0, 0.05)
  noise_life <- rnorm(n_countries, 0, 0.3)
  growth <- 1 + 0.02 + noise_gdp  # ~2% annual growth
  base_gdp <<- base_gdp * growth
  life_exp <- 45 + 8 * log(base_gdp / 1000) + noise_life
  pop <- abs(rnorm(n_countries, 5e6, 3e6)) * (1 + (yr - 2000) * 0.01)
  df <- bind_rows(df, tibble(
    country = paste0("Country_", 1:n_countries),
    year = yr,
    gdp = base_gdp,
    life_exp = pmin(life_exp, 85),
    pop = pop,
    continent = continents))
}

# ── 2. STATIC FRAME (2020) — the base plot ──────────────
p_static <- df |>
  filter(year == 2020) |>
  ggplot(aes(x = gdp, y = life_exp, size = pop, color = continent)) +
  geom_point(alpha = 0.6) +
  scale_x_log10(labels = scales::comma) +
  scale_size(range = c(2, 14), guide = "none") +
  scale_color_manual(values = c(
    Europe = "#1565C0", Asia = "#E53935",
    Africa = "#2E7D32", Americas = "#E65100")) +
  theme_minimal() +
  labs(title = "Gapminder-Style Static Frame: 2020",
       subtitle = "x = GDP/cap (log), y = life expectancy, size = population",
       x = "GDP per Capita (log scale)", y = "Life Expectancy",
       color = "Continent")
ggsave("output/gapminder_static_2020.png", p_static, width = 9, height = 6, dpi = 300)

# ── 3. FOUR KEY FRAMES (small multiples alternative) ────
key_years <- c(2000, 2007, 2014, 2020)
p_keyframes <- df |>
  filter(year %in% key_years) |>
  ggplot(aes(x = gdp, y = life_exp, size = pop, color = continent)) +
  geom_point(alpha = 0.5) +
  scale_x_log10(labels = scales::comma) +
  scale_size(range = c(1, 10), guide = "none") +
  scale_color_manual(values = c(
    Europe = "#1565C0", Asia = "#E53935",
    Africa = "#2E7D32", Americas = "#E65100")) +
  facet_wrap(~year, ncol = 4) +
  theme_minimal(base_size = 8) +
  labs(title = "Small Multiples: Static Alternative to Animation",
       subtitle = "Same data shown at 4 time points — all visible simultaneously",
       x = "GDP/cap (log)", y = "Life Exp.")
ggsave("output/gapminder_keyframes.png", p_keyframes, width = 16, height = 5, dpi = 300)

# ── 4. LINE REVEAL ANIMATION (transition_reveal) ────────
# Simulated Netflix-style cumulative data
nf_sim <- tibble(
  date = seq(as.Date("2015-01-01"), as.Date("2021-12-01"), by = "month"),
  titles = cumsum(rpois(84, 120)))

p_line_static <- ggplot(nf_sim, aes(x = date, y = titles)) +
  geom_line(color = "#1565C0", linewidth = 1.2) +
  geom_area(fill = "#1565C0", alpha = 0.08) +
  theme_minimal() +
  labs(title = "Cumulative Titles Over Time (static base for reveal animation)",
       x = NULL, y = "Cumulative Titles")
ggsave("output/line_reveal_static.png", p_line_static, width = 9, height = 4, dpi = 300)

# ANIMATED VERSION (uncomment to run — requires gganimate + gifski)
# library(gganimate)
#
# ── transition_reveal: line grows left-to-right ──
# p_reveal <- ggplot(nf_sim, aes(x = date, y = titles)) +
#   geom_line(color = "#1565C0", linewidth = 1.2) +
#   geom_area(fill = "#1565C0", alpha = 0.08) +
#   geom_point(color = "#E53935", size = 3) +  # dot at growing tip
#   theme_minimal() +
#   transition_reveal(date) +
#   labs(title = "Netflix Cumulative Titles",
#        subtitle = "Date: {frame_along}",
#        x = NULL, y = "Cumulative")
# animate(p_reveal, nframes = 84, fps = 10, width = 800, height = 400,
#   renderer = gifski_renderer("output/line_reveal.gif"))

# ── 5. gganimate: GAPMINDER (transition_time) ───────────
# p_gap <- df |>
#   ggplot(aes(x = gdp, y = life_exp, size = pop, color = continent)) +
#   geom_point(alpha = 0.6) +
#   scale_x_log10(labels = scales::comma) +
#   scale_size(range = c(2, 14), guide = "none") +
#   scale_color_manual(values = c(
#     Europe = "#1565C0", Asia = "#E53935",
#     Africa = "#2E7D32", Americas = "#E65100")) +
#   theme_minimal() +
#   # --- Animation layers ---
#   transition_time(year) +
#   labs(title = "Year: {frame_time}",
#        x = "GDP per Capita (log)", y = "Life Expectancy") +
#   ease_aes("cubic-in-out") +
#   shadow_wake(wake_length = 0.08, alpha = 0.2)
#
# animate(p_gap,
#   nframes = 120,  # total frames
#   fps = 10,       # frames per second
#   width = 800,
#   height = 600,
#   renderer = gifski_renderer("output/gapminder.gif"))

# ── 6. gganimate TRANSITIONS CHEATSHEET ─────────────────
# transition_time(year)      → continuous time (Gapminder)
# transition_states(year)    → discrete states (bar chart race)
# transition_reveal(date)    → cumulative reveal (line growing)
# transition_filter(cond)    → show/hide based on logical filter
# transition_layers()        → add layers sequentially
#
# EASING:
# ease_aes("linear")         → constant speed
# ease_aes("cubic-in-out")   → smooth acceleration/deceleration
# ease_aes("bounce-out")     → bouncing effect at end
#
# SHADOWS:
# shadow_wake(wake_length=)  → fading trail behind points
# shadow_mark(past=TRUE)     → keep all previous positions
# shadow_trail()             → dot trail following trajectory
#
# RENDERING:
# gifski_renderer()          → GIF (recommended)
# ffmpeg_renderer()          → MP4/video
# av_renderer()              → MP4 via av package

# ── 7. WHEN TO ANIMATE: DECISION FRAMEWORK ──────────────
# Create a visual reference card
decision <- tibble(
  scenario = c(
    "Show trajectory over time",
    "Presentation / keynote",
    "Social media engagement",
    "Compare exact values",
    "PDF report / journal paper",
    "Dashboard for analysis",
    "Gapminder exploration",
    "Print poster"),
  recommendation = c(
    "ANIMATE", "ANIMATE", "ANIMATE",
    "STATIC (small multiples)", "STATIC (key frames)",
    "STATIC (sparklines)", "ANIMATE", "STATIC"),
  reason = c(
    "Movement IS the story",
    "Audience captive, sequential",
    "Engagement, shareability",
    "Can't freeze animation frames",
    "Media doesn't support GIF",
    "Need random access, not sequence",
    "Trajectory + wake = insight",
    "No animation on paper"))

p_decision <- ggplot(decision,
  aes(x = fct_rev(fct_inorder(scenario)),
      y = 1, fill = recommendation)) +
  geom_tile(color = "white", linewidth = 1.5, height = 0.8) +
  geom_text(aes(label = paste0(recommendation, "\n", reason)),
    size = 2.5, lineheight = 1.1) +
  scale_fill_manual(values = c(
    ANIMATE = "#E3F2FD",
    "STATIC (small multiples)" = "#FFF3E0",
    "STATIC (key frames)" = "#FFF3E0",
    "STATIC (sparklines)" = "#FFF3E0",
    STATIC = "#FFF3E0")) +
  coord_flip() +
  theme_void(base_size = 9) +
  theme(axis.text.y = element_text(hjust = 1, face = "bold", size = 8),
        legend.position = "none") +
  labs(title = "Animation Decision Framework: When to Animate vs Static")
ggsave("output/animate_decision.png", p_decision, width = 10, height = 5, dpi = 300)

# ── 8. FRAME SEQUENCE (simulating animation as stills) ──
# Show the "animation" as a sequence of 6 stills
years_seq <- c(2000, 2004, 2008, 2012, 2016, 2020)
p_sequence <- df |>
  filter(year %in% years_seq) |>
  ggplot(aes(x = gdp, y = life_exp, size = pop, color = continent)) +
  geom_point(alpha = 0.5) +
  scale_x_log10() + scale_size(range = c(1, 8), guide = "none") +
  scale_color_manual(values = c(
    Europe = "#1565C0", Asia = "#E53935",
    Africa = "#2E7D32", Americas = "#E65100")) +
  facet_wrap(~year, ncol = 6) +
  theme_minimal(base_size = 7) +
  theme(legend.position = "bottom", legend.text = element_text(size = 6)) +
  labs(title = "Animation as 6 Stills: the frames gganimate interpolates between",
       x = "GDP/cap (log)", y = "Life Exp.")
ggsave("output/frame_sequence.png", p_sequence, width = 16, height = 4, dpi = 300)

cat("\n── All W06-M06 R outputs saved ──\n")
cat("Figures: gapminder_static_2020, gapminder_keyframes,\n")
cat("  line_reveal_static, animate_decision, frame_sequence\n")
cat("Animation code: commented out (requires gganimate + gifski)\n")

File: latest/week2/workshop6_m6/W06_M06_demo.py
"""W06-M06: Animation Fundamentals — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.animation import FuncAnimation
os.makedirs("output", exist_ok=True); np.random.seed(42)

# ── 1. SIMULATED GAPMINDER-STYLE DATA ───────────────────
n_countries = 40
continents = np.random.choice(["Europe","Asia","Africa","Americas"],
    n_countries, p=[0.25,0.30,0.25,0.20])
colors_map = {"Europe":"#1565C0","Asia":"#E53935","Africa":"#2E7D32","Americas":"#E65100"}

base_gdp = np.exp(np.random.uniform(6, 10, n_countries))
records = []
for yr in range(2000, 2021):
    base_gdp *= (1 + 0.02 + np.random.normal(0, 0.05, n_countries))
    life_exp = np.clip(45 + 8 * np.log(base_gdp / 1000) +
        np.random.normal(0, 0.3, n_countries), 30, 85)
    pop = np.abs(np.random.normal(5e6, 3e6, n_countries)) * (1 + (yr-2000)*0.01)
    for i in range(n_countries):
        records.append({"country": f"C{i}", "year": yr,
            "gdp": base_gdp[i], "life_exp": life_exp[i],
            "pop": pop[i], "continent": continents[i]})
df = pd.DataFrame(records)

# ── 2. STATIC FRAME — 2020 ──────────────────────────────
fr = df.query("year == 2020")
fig, ax = plt.subplots(figsize=(9, 6))
for cont, color in colors_map.items():
    sub = fr.query("continent == @cont")
    ax.scatter(sub["gdp"], sub["life_exp"], s=sub["pop"]/5e4,
        c=color, alpha=0.6, edgecolors="white", lw=0.5, label=cont)
ax.set_xscale("log")
ax.legend(fontsize=8)
ax.set_title("Gapminder-Style: 2020 (static base for animation)",
    fontweight="bold")
ax.set_xlabel("GDP per Capita (log)"); ax.set_ylabel("Life Expectancy")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/gapminder_static_2020.png", dpi=300); plt.close()

# ── 3. KEY FRAMES — 4-panel static alternative ──────────
fig, axes = plt.subplots(1, 4, figsize=(16, 4), sharex=True, sharey=True)
for ax, yr in zip(axes, [2000, 2007, 2014, 2020]):
    fr = df.query("year == @yr")
    for cont, color in colors_map.items():
        sub = fr.query("continent == @cont")
        ax.scatter(sub["gdp"], sub["life_exp"], s=sub["pop"]/8e4,
            c=color, alpha=0.5, edgecolors="white", lw=0.3)
    ax.set_xscale("log"); ax.set_title(str(yr), fontsize=11, fontweight="bold")
    ax.tick_params(labelsize=6)
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
axes[0].set_ylabel("Life Expectancy", fontsize=9)
fig.suptitle("Small Multiples: Static Alternative to Animation\n"
    "All 4 time points visible simultaneously", fontsize=12, fontweight="bold")
plt.tight_layout(); plt.savefig("output/gapminder_keyframes.png", dpi=300); plt.close()

# ── 4. CUMULATIVE LINE ANIMATION (FuncAnimation) ────────
months = pd.date_range("2015-01", "2021-12", freq="ME")
titles = np.cumsum(np.random.poisson(120, len(months)))

fig, ax = plt.subplots(figsize=(9, 4))
line, = ax.plot([], [], color="#1565C0", lw=1.5)
dot, = ax.plot([], [], "o", color="#E53935", markersize=6)
fill = None
ax.set_xlim(months[0], months[-1])
ax.set_ylim(0, titles[-1] * 1.1)
ax.set_title("Cumulative Line: transition_reveal() equivalent", fontweight="bold")
ax.set_ylabel("Cumulative Titles")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
from matplotlib.dates import DateFormatter
ax.xaxis.set_major_formatter(DateFormatter("%Y"))
title_text = ax.text(0.02, 0.95, "", transform=ax.transAxes,
    fontsize=10, fontweight="bold", va="top")

def init():
    line.set_data([], []); dot.set_data([], [])
    title_text.set_text(""); return line, dot, title_text

def update(frame):
    global fill
    if fill is not None:
        fill.remove()
    x = months[:frame+1]; y = titles[:frame+1]
    line.set_data(x, y)
    dot.set_data([x[-1]], [y[-1]])
    fill = ax.fill_between(x, y, alpha=0.06, color="#1565C0")
    title_text.set_text(f"{x[-1].strftime('%b %Y')} — {y[-1]:,} titles")
    return line, dot, title_text

anim = FuncAnimation(fig, update, frames=len(months),
    init_func=init, interval=80, blit=False)
try:
    anim.save("output/line_reveal.gif", writer="pillow", fps=12)
    print("Line reveal animation saved")
except Exception as e:
    print(f"Could not save line GIF: {e}")
plt.close()

# Static version showing 4 progress stages
fig, ax = plt.subplots(figsize=(9, 4))
for i, frac in enumerate([0.25, 0.50, 0.75, 1.0]):
    n = int(len(months) * frac)
    alpha = 0.2 if frac < 1.0 else 1.0
    lw = 0.8 if frac < 1.0 else 2.0
    label = f"{months[n-1].strftime('%b %Y')}" if frac == 1.0 else None
    ax.plot(months[:n], titles[:n], color="#1565C0", lw=lw, alpha=alpha, label=label)
    ax.scatter([months[n-1]], [titles[n-1]], s=30, c="#E53935", zorder=5)
ax.set_title("Line Reveal: 4 Progress Stages (static representation)",
    fontweight="bold")
ax.set_ylabel("Cumulative Titles"); ax.legend(fontsize=8)
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.xaxis.set_major_formatter(DateFormatter("%Y"))
plt.tight_layout(); plt.savefig("output/line_reveal_stages.png", dpi=300); plt.close()

# ── 5. GAPMINDER ANIMATION (FuncAnimation) ──────────────
years = list(range(2000, 2021))
fig, ax = plt.subplots(figsize=(9, 6))

def update_gap(frame_idx):
    ax.clear()
    yr = years[frame_idx]
    fr = df.query("year == @yr")
    for cont, color in colors_map.items():
        sub = fr.query("continent == @cont")
        ax.scatter(sub["gdp"], sub["life_exp"], s=sub["pop"]/5e4,
            c=color, alpha=0.6, edgecolors="white", lw=0.5, label=cont)
    ax.set_xscale("log")
    ax.set_xlim(200, 80000); ax.set_ylim(35, 90)
    ax.set_title(f"Year: {yr}", fontsize=14, fontweight="bold")
    ax.set_xlabel("GDP per Capita (log)"); ax.set_ylabel("Life Expectancy")
    ax.legend(fontsize=7, loc="lower right")
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
    # Year watermark
    ax.text(0.95, 0.05, str(yr), transform=ax.transAxes, fontsize=40,
        fontweight="bold", color="#EEEEEE", ha="right", va="bottom")
    return []

anim_gap = FuncAnimation(fig, update_gap, frames=len(years),
    interval=300, blit=False, repeat=True)
try:
    anim_gap.save("output/gapminder.gif", writer="pillow", fps=5)
    print("Gapminder animation saved")
except Exception as e:
    print(f"Could not save Gapminder GIF: {e}")
plt.close()

# ── 6. 6-FRAME SEQUENCE (animation as stills) ───────────
fig, axes = plt.subplots(1, 6, figsize=(18, 3.5), sharex=True, sharey=True)
for ax, yr in zip(axes, [2000, 2004, 2008, 2012, 2016, 2020]):
    fr = df.query("year == @yr")
    for cont, color in colors_map.items():
        sub = fr.query("continent == @cont")
        ax.scatter(sub["gdp"], sub["life_exp"], s=sub["pop"]/1e5,
            c=color, alpha=0.5, edgecolors="white", lw=0.3)
    ax.set_xscale("log"); ax.set_title(str(yr), fontsize=10, fontweight="bold")
    ax.tick_params(labelsize=5)
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Animation as 6 Stills: frames FuncAnimation interpolates between",
    fontsize=12, fontweight="bold")
plt.tight_layout(); plt.savefig("output/frame_sequence.png", dpi=300); plt.close()

# ── 7. FuncAnimation CHEATSHEET ──────────────────────────
print("\n=== FuncAnimation Cheatsheet ===")
print("fig, ax = plt.subplots()")
print("def init(): ...          # initialise empty artists")
print("def update(frame): ...   # redraw for each frame")
print("anim = FuncAnimation(fig, update,")
print("    frames=N,            # total frames")
print("    init_func=init,")
print("    interval=50,         # ms between frames")
print("    blit=True)           # fast rendering")
print("")
print("Saving:")
print("  anim.save('f.gif', writer='pillow', fps=10)")
print("  anim.save('f.mp4', writer='ffmpeg', fps=15)")
print("")
print("=== gganimate Equivalents ===")
print("transition_time(year)   → FuncAnimation with year loop")
print("transition_reveal(date) → cumulative line update")
print("shadow_wake()           → draw fading trail manually")
print("ease_aes('cubic')       → scipy.interpolate for smooth")

print("\nAll W06-M06 Python outputs saved")
